File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.py
def solution(friends, gifts):
    n = len(friends)
    idx = {f: i for i, f in enumerate(friends)}
    graph = [[0] * n for _ in range(n)]
    
    for gift in gifts:
        a, b = gift.split(' ')
        graph[idx[a]][idx[b]] += 1
        
    scores = [0] * n
    for i in range(n):
        give_score = sum(graph[i])
        receive_score = sum(graph[x][i] for x in range(n))
        scores[i] = give_score - receive_score
    
    counts = [0] * n
    for i in range(n):
        for j in range(n):
            if graph[i][j] > graph[j][i] or graph[i][j] == graph[j][i] and scores[i] > scores[j]:
                counts[i] += 1
    
    return max(counts)

File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.java
import java.util.*;

class Solution {

    public int solution(String[] friends, String[] gifts) {
        int n = friends.length;
        Map<String, Integer> nameToIndex = new HashMap<>();
        for (int i = 0; i < n; i++) {
            nameToIndex.put(friends[i], i);
        }

        int[][] giftGraph = new int[n][n];
        int[] giftScore = new int[n];

        for (String gift : gifts) {
            String[] participants = gift.split(" ");
            int giver = nameToIndex.get(participants[0]);
            int receiver = nameToIndex.get(participants[1]);
            
            giftGraph[giver][receiver]++;
            giftScore[giver]++;
            giftScore[receiver]--;
        }

        int[] nextMonthGifts = new int[n];
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                int iToJ = giftGraph[i][j];
                int jToI = giftGraph[j][i];

                if (iToJ > jToI) {
                    nextMonthGifts[i]++;
                } else if (iToJ < jToI) {
                    nextMonthGifts[j]++;
                } else {
                    if (giftScore[i] > giftScore[j]) {
                        nextMonthGifts[i]++;
                    } else if (giftScore[i] < giftScore[j]) {
                        nextMonthGifts[j]++;
                    }
                }
            }
        }

        return Arrays.stream(nextMonthGifts).max().orElse(0);
    }
}